### PR TITLE
[AIRFLOW-2613] Fix Airflow searching .zip bug

### PR DESCRIPTION
**Pull request by Noremac201**
...and add DAG skip logging

### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-2613

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:
When Airflow was populating a DagBag from a .zip file, if a single
file in the root directory did not contain the strings 'airflow' and
'DAG' it would ignore the entire .zip file.

Also added a small amount of logging to not bombard user with info
about skipping their .py files.
### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Adds a file in the test_zip where a .py file does not contain "airflow" nor "DAG". This checks to make sure the entire .zip is not skipped over.

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### Code Quality
- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

**Comment by feng-tao** [comment]
now after this change,  i wonder how do we identify if a file is a DAG file. for example, if a file without "DAG", do we still consider a DAG file? If I understand correctly, it will import any python file under the dagbag directory even without the DAG keyword?

I guess Max @mistercrunch  may have a better idea on why we need this in first place.

**Comment by artwr** [comment]
The goal in this case was to prevent the DagBag parsing from importing all python files people were collocating with their DAGs.

**Comment by feng-tao** [comment]
thanks @artwr for the info. Then how do we prevent import any arbitrary python files in the dagbag directory after this pr?

**Comment by artwr** [comment]
@feng-tao there is no easy way without parsing and executing the code to know whether a Python file loads a DAG object into its globals. The regex/string matching version is a cheap alternative to actual parsing + execution.

**Comment by artwr** [comment]
I should say because of this I am a -1 on this for now, because I know for a fact that this might mess up the DagBags for a few people.

**Comment by TrevorEdwards** [comment]
Can you explain how this will mess up DagBags for people? In the worst case, this could increase processing time for people who put many non-dag Python files in the top-level directory; however, I don't see how this would cause incorrect execution except in exceptional circumstances.

There is a .airflowignore file supported if you want to skip specific files.

On the other hand, it is counter-intuitive for processing of a zip file to stop simply because one file in it does not have the words "DAG" and "airflow". It's even worse for this to happen silently.

**Comment by artwr** [comment]
>> Can you explain how this will mess up DagBags for people? In the worst case, this could increase processing time for people who put many non-dag Python files in the top-level directory;

Sorry, it was more of a drive by comment, by mess up, I meant operationally, as it might result in a larger number of files parsed, but not that it would prevent Dags from being loaded. Some colocated python scripts could call an expensive resource (and we have seen this in the past). It also not only the top level directory since we are doing an os.walk in collect_dags.

I agree that the check probably needs to be refined, but it does not seem completely crazy that the words airflow/DAG should be in a module creating a DAG, if only in a docstring at the top of the file. It is definitely good to document/we could expand the actual check though to more than just those two strings. 

`.airflowignore` is fine, but it does not address the issue at hand as it is more of a dev facility than a production one. You would have to add all the non dag files which you want to ignore there. What happens if someone has a DAG file named the same as the script you are trying to exclude...?

Complete agreed with you on the zipfile front, the check should probably look into all files, and if any are found, return. I would welcome a PR which would make that change.

**Comment by TrevorEdwards** [comment]
Ah okay. I'm not sure of an easy way to detect with full confidence if a file has a DAG in its globals without interpreting it. But it is reasonable that we would want to avoid breaking / slowing down existing workloads.

For context, it took me a few hours to debug why a few DAGs I had just written were not being loaded, but if the scheduler had at least logged that it skipped a file due to it not containing those strings, it would have saved a lot of time. I'd suggest logging the fact that a file was skipped, but I think we should at least document it if you disagree with logging.

**Comment by Noremac201** [comment]
After taking all of your comments into consideration, I have updated it to include the following.
- Airflow will log that it believes there is no DAG in a file, once per DagBag. This is to not spam users with a ton of skip messages.
- I've updated docs mentioning in two places how Airflow's logic works, so it is much more apparent how it skips the files.
- If a single file does not contain "airflow" and "DAG" in a zip file, it will not return immediately. Thus allowing to load all DAGs from a .zip.

Let me know if there are any changes you'd like to see.
Thanks for your feedback!

File: docs/concepts.rst [hunk 1]
```diff
@@ -35,6 +35,9 @@ the ``DAG`` objects. You can have as many DAGs as you want, each describing an
 arbitrary number of tasks. In general, each one should correspond to a single
 logical workflow.
 
+.. note:: When searching for DAGs, Airflow will only consider files where the string
+   "airflow" and "DAG" both appear in the contents of the ``.py`` file.
```
Comment (artwr) [inline docs/concepts.rst:39]: Thank you!

**Review by artwr — APPROVED** [review]
This LGTM now

**Comment by feng-tao** [comment]
is the zip file in test folder still needed?

**Comment by artwr** [comment]
@TrevorEdwards @Noremac201 The long term solution is most likely the DAGFetcher abstraction, which would know a little better how to find DAGs, either by using a manifest (Basically, you add the relevant file to a list) or another litmus test for the existence of a DAG. By making this configurable, we could allow users to define what is right for their infrastructure, with a default out of the box to the old behaviour.

**Comment by Noremac201** [comment]
@feng-tao I believe so, there's currently a unit test that uses it to test getting a DAG from inside of a zip file. All I did was add another .py in the .zip which doesn't contain "airflow" or "DAG" to make sure it doesn't skip DAGs like before.

File: airflow/models.py [hunk 2]
```diff
@@ -219,6 +219,7 @@ def __init__(
         self.file_last_changed = {}
         self.executor = executor
         self.import_errors = {}
+        self.has_logged = False
```
Comment (kaxil) [inline airflow/models.py:226]: Can you add a docstring for this param?

https://github.com/Noremac201/incubator-airflow/blob/b271cab470bfebc7e46aae0377f68dabe3e03c5f/airflow/models.py#L191-L198

File: airflow/models.py [hunk 3]
```diff
@@ -333,7 +340,12 @@ def process_file(self, filepath, only_if_updated=True, safe_mode=True):
                                 self.file_last_changed[filepath] = (
                                     file_last_changed_on_disk)
                                 # todo: create ignore list
-                                return found_dags
+                                # Don't want to spam user with skip messages
+                                if not self.has_logged:
+                                    self.has_logged = True
+                                    self.log.info(
+                                        "File %s assumed to contain no DAGs. Skipping.",
```
Comment (kaxil) [inline airflow/models.py:351]: @Noremac201 Can you add a test to see if logs this message?

**Review by kaxil — CHANGES_REQUESTED** [review]

**Comment by kaxil** [comment]
Also, can you update the PR title as well to match your commit title?

File: airflow/models.py (hunk 2 above)
Comment (Noremac201) [reply airflow/models.py:226]: has_logged isn't passed to the constructor, do you still want me to document it?

Comment (kaxil) [reply airflow/models.py:226]: I know but the param name itself isn't self-explanatory so I am saying so.

**Review by kaxil — APPROVED** [review]
LGTM

**Pull request closed without merging**